File: scripts/omnicontext_wrapper.py
#!/usr/bin/env python3
"""
Programmatic wrapper for OmniContext MCP server.

This script provides controlled access to omnicontext tools without
automatically injecting full response bodies into context. You can:
- Filter and summarize responses
- Chain multiple tool calls
- Extract only relevant information
- Control token usage precisely
"""

import json
import subprocess
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional


class OmniContextWrapper:
    """Wrapper for OmniContext MCP server with controlled output."""
    
    def __init__(self, repo_path: str, mcp_exe_path: Optional[str] = None):
        self.repo_path = Path(repo_path).resolve()
        
        # Auto-detect MCP executable
        if mcp_exe_path:
            self.mcp_exe = Path(mcp_exe_path)
        else:
            # Try common locations
            home = Path.home()
            candidates = [
                home / ".omnicontext" / "bin" / "omnicontext-mcp.exe",
                home / ".omnicontext" / "bin" / "omnicontext-mcp",
                Path("target/release/omnicontext-mcp.exe"),
                Path("target/release/omnicontext-mcp"),
            ]
            self.mcp_exe = next((p for p in candidates if p.exists()), None)
            
        if not self.mcp_exe or not self.mcp_exe.exists():
            raise FileNotFoundError(f"OmniContext MCP executable not found")
    
    def _call_tool(self, tool_name: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """Call an MCP tool and return raw response."""
        cmd = [str(self.mcp_exe), "--repo", str(self.repo_path)]
        
        # Create MCP request
        request = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "tools/call",
            "params": {
                "name": tool_name,
                "arguments": params
            }
        }
        
        try:
            result = subprocess.run(
                cmd,
                input=json.dumps(request),
                capture_output=True,
                text=True,
                timeout=30
            )
            
            if result.returncode != 0:
                return {"error": f"Process failed: {result.stderr}"}
            
            response = json.loads(result.stdout)
            return response.get("result", {})
        except subprocess.TimeoutExpired:
            return {"error": "Tool call timed out"}
        except json.JSONDecodeError as e:
            return {"error": f"Invalid JSON response: {e}"}
        except Exception as e:
            return {"error": f"Tool call failed: {e}"}
    
    # ===== High-level API with controlled output =====
    
    def get_status_summary(self) -> str:
        """Get concise status summary (< 200 tokens)."""
        result = self._call_tool("get_status", {})
        
        if "error" in result:
            return f"Error: {result['error']}"
        
        # Extract only key metrics
        content = result.get("content", [{}])[0].get("text", "")
        lines = content.split("\n")
        
        summary = []
        for line in lines:
            if any(k in line for k in ["Files:", "Chunks:", "Symbols:", "Vectors:", "Search mode:"]):
                summary.append(line.strip())
        
        return "\n".join(summary[:5])  # Max 5 lines
    
    def search_code_filtered(
        self, 
        query: str, 
        limit: int = 5,
        max_tokens_per_result: int = 200
    ) -> List[Dict[str, str]]:
        """Search code and return filtered results."""
        result = self._call_tool("search_code", {"query": query, "limit": limit})
        
        if "error" in result:
            return [{"error": result["error"]}]
        
        content = result.get("content", [{}])[0].get("text", "")
        
        # Parse and filter results
        filtered = []
        current_result = {}
        
        for line in content.split("\n"):
            if line.startswith("**") and "**" in line[2:]:
                # New result
                if current_result:
                    filtered.append(current_result)
                current_result = {"title": line.strip("* ")}
            elif ":" in line and current_result:
                key, value = line.split(":", 1)
                current_result[key.strip().lower()] = value.strip()
            elif line.strip().startswith("```") and current_result:
                # Truncate code snippets
                current_result["code_preview"] = "[code truncated]"
        
        if current_result:
            filtered.append(current_result)
        
        return filtered[:limit]
    
    def get_symbol_info(self, symbol_name: str) -> Dict[str, Any]:
        """Get symbol info with minimal context."""
        result = self._call_tool("get_symbol", {"name": symbol_name, "limit": 3})
        
        if "error" in result:
            return {"error": result["error"]}
        
        content = result.get("content", [{}])[0].get("text", "")
        
        # Extract just the symbol list, not full code
        symbols = []
        for line in content.split("\n"):
            if line.startswith("- **"):
                symbols.append(line.strip("- "))
        
        return {
            "query": symbol_name,
            "found": len(symbols),
            "symbols": symbols[:3]  # Max 3 results
        }
    
    def get_architecture_summary(self) -> str:
        """Get high-level architecture (< 500 tokens)."""
        result = self._call_tool("get_architecture", {})
        
        if "error" in result:
            return f"Error: {result['error']}"
        
        content = result.get("content", [{}])[0].get("text", "")
        
        # Extract only summary sections, skip detailed listings
        lines = content.split("\n")
        summary = []
        skip_section = False
        
        for line in lines:
            if line.startswith("##"):
                skip_section = "Indexed Content" in line or "Recommendations" in line
            
            if not skip_section and (line.startswith("#") or line.startswith("-") or ":" in line):
                summary.append(line)
        
        return "\n".join(summary[:20])  # Max 20 lines
    
    def find_patterns_summary(self, pattern: str, limit: int = 3) -> str:
        """Find patterns with minimal output."""
        result = self._call_tool("find_patterns", {"pattern": pattern, "limit": limit})
        
        if "error" in result:
            return f"Error: {result['error']}"
        
        content = result.get("content", [{}])[0].get("text", "")
        
        # Return just file paths and line numbers, not full code
        summary = []
        for line in content.split("\n"):
            if "file:" in line.lower() or "line:" in line.lower():
                summary.append(line.strip())
        
        return "\n".join(summary[:10])  # Max 10 lines
    
    def get_dependencies_graph(self, symbol: str, direction: str = "both") -> Dict[str, List[str]]:
        """Get dependency graph in compact format."""
        result = self._call_tool("get_dependencies", {
            "symbol": symbol,
            "direction": direction
        })
        
        if "error" in result:
            return {"error": [result["error"]]}
        
        content = result.get("content", [{}])[0].get("text", "")
        
        # Parse into structured format
        deps = {"upstream": [], "downstream": []}
        current_section = None
        
        for line in content.split("\n"):
            if "Upstream" in line:
                current_section = "upstream"
            elif "Downstream" in line:
                current_section = "downstream"
            elif line.startswith("- ") and current_section:
                deps[current_section].append(line.strip("- `"))
        
        return deps
    
    def context_window_compact(
        self, 
        query: str, 
        token_budget: int = 2000
    ) -> str:
        """Get context window with strict token budget."""
        result = self._call_tool("context_window", {
            "query": query,
            "limit": 10,
            "token_budget": token_budget
        })
        
        if "error" in result:
            return f"Error: {result['error']}"
        
        content = result.get("content", [{}])[0].get("text", "")
        
        # Truncate to budget (rough estimate: 4 chars = 1 token)
        max_chars = token_budget * 4
        if len(content) > max_chars:
            content = content[:max_chars] + "\n\n[... truncated to fit token budget]"
        
        return content
    
    # ===== Chained operations =====
    
    def analyze_symbol_full(self, symbol_name: str) -> Dict[str, Any]:
        """Chain multiple calls to fully analyze a symbol."""
        analysis = {
            "symbol": symbol_name,
            "info": self.get_symbol_info(symbol_name),
            "dependencies": self.get_dependencies_graph(symbol_name),
        }
        
        # Only fetch context if symbol was found
        if analysis["info"].get("found", 0) > 0:
            analysis["context"] = self.context_window_compact(
                symbol_name, 
                token_budget=1000
            )
        
        return analysis
    
    def search_and_analyze(self, query: str, top_n: int = 3) -> Dict[str, Any]:
        """Search and analyze top results in one call."""
        results = self.search_code_filtered(query, limit=top_n)
        
        analysis = {
            "query": query,
            "results_found": len(results),
            "top_results": results,
        }
        
        # Get architecture context for the query
        analysis["architecture"] = self.get_architecture_summary()
        
        return analysis
    
    def health_check(self) -> Dict[str, Any]:
        """Quick health check of the index."""
        status = self.get_status_summary()
        
        # Parse key metrics
        metrics = {}
        for line in status.split("\n"):
            if ":" in line:
                key, value = line.split(":", 1)
                metrics[key.strip().lower()] = value.strip()
        
        return {
            "healthy": "error" not in status.lower(),
            "metrics": metrics,
            "raw_status": status
        }


def main():
    """CLI interface for the wrapper."""
    if len(sys.argv) < 3:
        print("Usage: omnicontext_wrapper.py <repo_path> <command> [args...]")
        print("\nCommands:")
        print("  status              - Get status summary")
        print("  search <query>      - Search code")
        print("  symbol <name>       - Get symbol info")
        print("  deps <symbol>       - Get dependencies")
        print("  analyze <symbol>    - Full symbol analysis")
        print("  health              - Health check")
        sys.exit(1)
    
    repo_path = sys.argv[1]
    command = sys.argv[2]
    
    wrapper = OmniContextWrapper(repo_path)
    
    if command == "status":
        print(wrapper.get_status_summary())
    
    elif command == "search" and len(sys.argv) > 3:
        query = " ".join(sys.argv[3:])
        results = wrapper.search_code_filtered(query)
        print(json.dumps(results, indent=2))
    
    elif command == "symbol" and len(sys.argv) > 3:
        symbol = sys.argv[3]
        info = wrapper.get_symbol_info(symbol)
        print(json.dumps(info, indent=2))
    
    elif command == "deps" and len(sys.argv) > 3:
        symbol = sys.argv[3]
        deps = wrapper.get_dependencies_graph(symbol)
        print(json.dumps(deps, indent=2))
    
    elif command == "analyze" and len(sys.argv) > 3:
        symbol = sys.argv[3]
        analysis = wrapper.analyze_symbol_full(symbol)
        print(json.dumps(analysis, indent=2))
    
    elif command == "health":
        health = wrapper.health_check()
        print(json.dumps(health, indent=2))
    
    else:
        print(f"Unknown command: {command}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/omnicontext_wrapper.ts
#!/usr/bin/env node
/**
 * Programmatic wrapper for OmniContext MCP server (TypeScript/Node.js version)
 * 
 * Provides controlled access to omnicontext tools without automatically
 * injecting full response bodies into context.
 */

import { spawn } from 'child_process';
import { existsSync } from 'fs';
import { homedir } from 'os';
import { join, resolve } from 'path';

interface MCPRequest {
    jsonrpc: string;
    id: number;
    method: string;
    params: {
        name: string;
        arguments: Record<string, any>;
    };
}

interface MCPResponse {
    result?: {
        content?: Array<{ text: string }>;
        error?: string;
    };
    error?: any;
}

interface SearchResult {
    title?: string;
    file?: string;
    line?: string;
    score?: string;
    code_preview?: string;
}

interface SymbolInfo {
    query: string;
    found: number;
    symbols: string[];
}

interface DependencyGraph {
    upstream: string[];
    downstream: string[];
}

export class OmniContextWrapper {
    private repoPath: string;
    private mcpExe: string;

    constructor(repoPath: string, mcpExePath?: string) {
        this.repoPath = resolve(repoPath);

        if (mcpExePath) {
            this.mcpExe = mcpExePath;
        } else {
            // Auto-detect MCP executable
            const home = homedir();
            const candidates = [
                join(home, '.omnicontext', 'bin', 'omnicontext-mcp.exe'),
                join(home, '.omnicontext', 'bin', 'omnicontext-mcp'),
                join('target', 'release', 'omnicontext-mcp.exe'),
                join('target', 'release', 'omnicontext-mcp'),
            ];

            const found = candidates.find(p => existsSync(p));
            if (!found) {
                throw new Error('OmniContext MCP executable not found');
            }
            this.mcpExe = found;
        }
    }

    private async callTool(toolName: string, params: Record<string, any>): Promise<any> {
        const request: MCPRequest = {
            jsonrpc: '2.0',
            id: 1,
            method: 'tools/call',
            params: {
                name: toolName,
                arguments: params,
            },
        };

        return new Promise((resolve, reject) => {
            const proc = spawn(this.mcpExe, ['--repo', this.repoPath]);

            let stdout = '';
            let stderr = '';

            proc.stdout.on('data', (data) => {
                stdout += data.toString();
            });

            proc.stderr.on('data', (data) => {
                stderr += data.toString();
            });

            proc.on('close', (code) => {
                if (code !== 0) {
                    reject(new Error(`Process failed: ${stderr}`));
                    return;
                }

                try {
                    const response: MCPResponse = JSON.parse(stdout);
                    resolve(response.result || {});
                } catch (e) {
                    reject(new Error(`Invalid JSON response: ${e}`));
                }
            });

            proc.on('error', (err) => {
                reject(err);
            });

            // Send request
            proc.stdin.write(JSON.stringify(request));
            proc.stdin.end();

            // Timeout after 30 seconds
            setTimeout(() => {
                proc.kill();
                reject(new Error('Tool call timed out'));
            }, 30000);
        });
    }

    /**
     * Get concise status summary (< 200 tokens)
     */
    async getStatusSummary(): Promise<string> {
        try {
            const result = await this.callTool('get_status', {});
            const content = result.content?.[0]?.text || '';

            const lines = content.split('\n');
            const summary = lines
                .filter(line =>
                    ['Files:', 'Chunks:', 'Symbols:', 'Vectors:', 'Search mode:']
                        .some(k => line.includes(k))
                )
                .slice(0, 5);

            return summary.join('\n');
        } catch (error) {
            return `Error: ${error}`;
        }
    }

    /**
     * Search code and return filtered results
     */
    async searchCodeFiltered(
        query: string,
        limit: number = 5,
        maxTokensPerResult: number = 200
    ): Promise<SearchResult[]> {
        try {
            const result = await this.callTool('search_code', { query, limit });
            const content = result.content?.[0]?.text || '';

            const filtered: SearchResult[] = [];
            let currentResult: SearchResult = {};

            for (const line of content.split('\n')) {
                if (line.startsWith('**') && line.includes('**', 2)) {
                    if (Object.keys(currentResult).length > 0) {
                        filtered.push(currentResult);
                    }
                    currentResult = { title: line.replace(/\*\*/g, '').trim() };
                } else if (line.includes(':') && Object.keys(currentResult).length > 0) {
                    const [key, ...valueParts] = line.split(':');
                    const value = valueParts.join(':').trim();
                    currentResult[key.trim().toLowerCase()] = value;
                } else if (line.trim().startsWith('```') && Object.keys(currentResult).length > 0) {
                    currentResult.code_preview = '[code truncated]';
                }
            }

            if (Object.keys(currentResult).length > 0) {
                filtered.push(currentResult);
            }

            return filtered.slice(0, limit);
        } catch (error) {
            return [{ title: `Error: ${error}` }];
        }
    }

    /**
     * Get symbol info with minimal context
     */
    async getSymbolInfo(symbolName: string): Promise<SymbolInfo> {
        try {
            const result = await this.callTool('get_symbol', { name: symbolName, limit: 3 });
            const content = result.content?.[0]?.text || '';

            const symbols = content
                .split('\n')
                .filter(line => line.startsWith('- **'))
                .map(line => line.replace(/^- \*\*/, '').replace(/\*\*.*$/, '').trim())
                .slice(0, 3);

            return {
                query: symbolName,
                found: symbols.length,
                symbols,
            };
        } catch (error) {
            return {
                query: symbolName,
                found: 0,
                symbols: [`Error: ${error}`],
            };
        }
    }

    /**
     * Get high-level architecture (< 500 tokens)
     */
    async getArchitectureSummary(): Promise<string> {
        try {
            const result = await this.callTool('get_architecture', {});
            const content = result.content?.[0]?.text || '';

            const lines = content.split('\n');
            const summary: string[] = [];
            let skipSection = false;

            for (const line of lines) {
                if (line.startsWith('##')) {
                    skipSection = line.includes('Indexed Content') || line.includes('Recommendations');
                }

                if (!skipSection && (line.startsWith('#') || line.startsWith('-') || line.includes(':'))) {
                    summary.push(line);
                }
            }

            return summary.slice(0, 20).join('\n');
        } catch (error) {
            return `Error: ${error}`;
        }
    }

    /**
     * Find patterns with minimal output
     */
    async findPatternsSummary(pattern: string, limit: number = 3): Promise<string> {
        try {
            const result = await this.callTool('find_patterns', { pattern, limit });
            const content = result.content?.[0]?.text || '';

            const summary = content
                .split('\n')
                .filter(line => line.toLowerCase().includes('file:') || line.toLowerCase().includes('line:'))
                .slice(0, 10);

            return summary.join('\n');
        } catch (error) {
            return `Error: ${error}`;
        }
    }

    /**
     * Get dependency graph in compact format
     */
    async getDependenciesGraph(symbol: string, direction: string = 'both'): Promise<DependencyGraph> {
        try {
            const result = await this.callTool('get_dependencies', { symbol, direction });
            const content = result.content?.[0]?.text || '';

            const deps: DependencyGraph = { upstream: [], downstream: [] };
            let currentSection: 'upstream' | 'downstream' | null = null;

            for (const line of content.split('\n')) {
                if (line.includes('Upstream')) {
                    currentSection = 'upstream';
                } else if (line.includes('Downstream')) {
                    currentSection = 'downstream';
                } else if (line.startsWith('- ') && currentSection) {
                    deps[currentSection].push(line.replace(/^- `?/, '').replace(/`?$/, '').trim());
                }
            }

            return deps;
        } catch (error) {
            return { upstream: [`Error: ${error}`], downstream: [] };
        }
    }

    /**
     * Get context window with strict token budget
     */
    async contextWindowCompact(query: string, tokenBudget: number = 2000): Promise<string> {
        try {
            const result = await this.callTool('context_window', {
                query,
                limit: 10,
                token_budget: tokenBudget,
            });

            let content = result.content?.[0]?.text || '';

            // Truncate to budget (rough estimate: 4 chars = 1 token)
            const maxChars = tokenBudget * 4;
            if (content.length > maxChars) {
                content = content.substring(0, maxChars) + '\n\n[... truncated to fit token budget]';
            }

            return content;
        } catch (error) {
            return `Error: ${error}`;
        }
    }

    /**
     * Chain multiple calls to fully analyze a symbol
     */
    async analyzeSymbolFull(symbolName: string): Promise<any> {
        const info = await this.getSymbolInfo(symbolName);
        const dependencies = await this.getDependenciesGraph(symbolName);

        const analysis: any = {
            symbol: symbolName,
            info,
            dependencies,
        };

        // Only fetch context if symbol was found
        if (info.found > 0) {
            analysis.context = await this.contextWindowCompact(symbolName, 1000);
        }

        return analysis;
    }

    /**
     * Search and analyze top results in one call
     */
    async searchAndAnalyze(query: string, topN: number = 3): Promise<any> {
        const results = await this.searchCodeFiltered(query, topN);
        const architecture = await this.getArchitectureSummary();

        return {
            query,
            results_found: results.length,
            top_results: results,
            architecture,
        };
    }

    /**
     * Quick health check of the index
     */
    async healthCheck(): Promise<any> {
        const status = await this.getStatusSummary();

        const metrics: Record<string, string> = {};
        for (const line of status.split('\n')) {
            if (line.includes(':')) {
                const [key, value] = line.split(':');
                metrics[key.trim().toLowerCase()] = value.trim();
            }
        }

        return {
            healthy: !status.toLowerCase().includes('error'),
            metrics,
            raw_status: status,
        };
    }
}

// CLI interface
async function main() {
    const args = process.argv.slice(2);

    if (args.length < 2) {
        console.log('Usage: omnicontext_wrapper.ts <repo_path> <command> [args...]');
        console.log('\nCommands:');
        console.log('  status              - Get status summary');
        console.log('  search <query>      - Search code');
        console.log('  symbol <name>       - Get symbol info');
        console.log('  deps <symbol>       - Get dependencies');
        console.log('  analyze <symbol>    - Full symbol analysis');
        console.log('  health              - Health check');
        process.exit(1);
    }

    const [repoPath, command, ...cmdArgs] = args;
    const wrapper = new OmniContextWrapper(repoPath);

    try {
        switch (command) {
            case 'status':
                console.log(await wrapper.getStatusSummary());
                break;

            case 'search':
                if (cmdArgs.length === 0) {
                    console.error('Error: search requires a query');
                    process.exit(1);
                }
                const searchResults = await wrapper.searchCodeFiltered(cmdArgs.join(' '));
                console.log(JSON.stringify(searchResults, null, 2));
                break;

            case 'symbol':
                if (cmdArgs.length === 0) {
                    console.error('Error: symbol requires a name');
                    process.exit(1);
                }
                const symbolInfo = await wrapper.getSymbolInfo(cmdArgs[0]);
                console.log(JSON.stringify(symbolInfo, null, 2));
                break;

            case 'deps':
                if (cmdArgs.length === 0) {
                    console.error('Error: deps requires a symbol');
                    process.exit(1);
                }
                const deps = await wrapper.getDependenciesGraph(cmdArgs[0]);
                console.log(JSON.stringify(deps, null, 2));
                break;

            case 'analyze':
                if (cmdArgs.length === 0) {
                    console.error('Error: analyze requires a symbol');
                    process.exit(1);
                }
                const analysis = await wrapper.analyzeSymbolFull(cmdArgs[0]);
                console.log(JSON.stringify(analysis, null, 2));
                break;

            case 'health':
                const health = await wrapper.healthCheck();
                console.log(JSON.stringify(health, null, 2));
                break;

            default:
                console.error(`Unknown command: ${command}`);
                process.exit(1);
        }
    } catch (error) {
        console.error(`Error: ${error}`);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}
